Export and predict

phildow opened the pull request:
Ok, so this one's a fatty, but I've been pushing on the code pretty hard to support export and prediction and been refactoring and documenting along the way. I'd like to drop https://github.com/doc-ai/aleph/pull/6 and replace it with this PR, as it includes all the code there.

Up next: tests tests tests and more error checking in the code.

nkashy1 commented:
@phildow : Could you add a comment summarizing the changes that this PR introduces and how we could test them?

nkashy1 commented:
For now, I am simply going to go through the README and make sure that everything works that the README says works. I will also look at the corresponding code as I go along and leave comments on this PR.

nkashy1 commented on README.md:
```diff
@@ -1,6 +1,596 @@
+
+The datasets, features, labels, models, and optimizers commands simply define project files. Datasets, features, and labels are nothing more than python dictionaries that are used later on. Models and optimizers are templated files that you will add custom code to. The train, export, and predict commands then read the datasets, features, and labels you have defined and  execute project specific * ...
+
+### 1.1 aleph init
```
From a user experience point of view, would it be possible to specify the working directory for the project one is initializing? I like to have my virtualenvs in my project directories and this init workflow makes that difficult (the virtualenv from which I run `aleph init` needs to already exist prior to my aleph project directory existing).

phildow replied:
Ah yeah, that's what I was trying to do with that earlier code, where the init command also bootstraps a virtual env in that directory, but then you still have to have aleph installed to do that, so a catch 22.

There could be some kind of argument passed to init like "here", or init without a name, `aleph init here` or `aleph init` that just initializes the project in the current working directory.

nkashy1 replied:
Yeah `aleph init .` sounds right -- it just should not complain then if the directory already exists (only on `.`?)

nkashy1 replied:
That can be an issue.

phildow replied:
Exactly. On `aleph init .` just scaffold the project in the cwd otherwise behave as it currently does.

phildow replied:
https://github.com/doc-ai/aleph/issues/25

phildow commented:
Major changes:

- Support the aleph train, aleph export, and aleph predict commands
- Renamed the aleph.aleph_cli module to aleph.cli
- Added support for train, export, and predict parameterization
- Most of the commits are small refactorings, moving code around mostly
- Added REAME

I'm wrapping up some parameterization code in a local branch. Then with the readme done I'll be shifting focus to testing.

For understanding the project: if you look at the code directly rather than at the PRs, every submodule in the aleph.cli module follows the same structure: 

1. Each subcommand belongs in its own submodule under aleph.cli, e.g. aleph.cli.datasets, aleph.cli.features, etc.
2. Each submodule sets up an argparse subparser in arg_parser.py
3. Each submodule routes incoming commands to a handler method when that subcommand is called. This occurs in generator.py.
4. Most submodules have templates that the generator might copy to the project when it executes a subcommand.
5. A number of submodules have a context.py file which is responsible for preparing content for the params dictionary that is ultimately passed to a project's train.py, export.py, and predict.py files. Those methods are called from the train, export, and predict submodules.

The global entry point to all the submodules is aleph.command_line_interface.

Testing can actually follow the example you established with aleph.cli.init.test_generator. Prepare an args dictionary and pass it to the generator method, confirming that it modifies the right project files or calls the right project methods, in the case of train, export, and predict.

I do want to move testing out of the aleph.cli and aleph.project modules into its own submodule, following the directory structure here:

https://stackoverflow.com/questions/1896918/running-unittest-with-typical-test-directory-structure

nkashy1 commented on README.md:
```diff
@@ -1,6 +1,596 @@
+
+The training script Aleph uses to train models. You may want to make modifications to this file.
+
+### 1.2 aleph datasets
```
It's pretty hard to test all these features from scratch.

@phildow : Could you prioritize adding a sample dataset (very small) that we can use to run an end-to-end test manually (and which we could later automate into an automated test)?

phildow replied:
Definitely. I plan to add an entire sample project. Maybe use a subset of the cats-vs-dogs dataset, make it easy to run all this stuff?

nkashy1 replied:
Exactly. An end-to-end automated test of the workflow is what we need (including training, prediction, etc.). No need to do it all in one go, though.

phildow replied:
https://github.com/doc-ai/aleph/issues/15

nkashy1 commented on README.md:
````diff
@@ -1,6 +1,596 @@
+Define a model for your project with the aleph models add command. Its signature looks like:
+
+```
+aleph models add name
````
It makes sense to connect a model to features and labels. Once we have more sophisticated code generation, we can generate a model-appropriate input function.

phildow replied:
Not sure what you mean. The way I understood this is that you might build completely different models and want to try them on the same dataset.

The problem with code generation is that we can't guarantee the order that the user runs .the commands in, and if we try to re-generate previously generated code, we might overwrite code the user has added.

Image I go:

```
aleph features add ...
aleph models add ...  => code generates on that feature
aleph features add 
```

nkashy1 replied:
@phildow : Something like this:
```
aleph models add inception-v3 labels=height,weight,age,gender features=image
```

This allows for different versions of datasets, slightly differently formatted datasets, etc.

phildow replied:
can we make this a longer running issue? i feel like this gets complex very quickly

phildow replied:
https://github.com/doc-ai/aleph/issues/27

nkashy1 commented on README.md:
````diff
@@ -1,6 +1,596 @@
+Define an optimizer for your project with the aleph optimizers add command. Its signature looks like:
+
+```
+aleph optimizers add name type
````
Along with optimizers, we should also allow users to specify a loss function. Standard loss functions (cross entropy, mean squared error, etc.) or user-defined losses as we have used internally. Ideally, this would generate a $LOSS_NAME.py file which would get imported into the model file.

phildow replied:
I remember we spoke about this, but I disagree. Or rather, let me ask this. Given some model x, does it make sense to try different loss functions with that model, in the way it makes sense to try different optimizers with that model? I think you're suggesting it does, even for more sophisticated loss functions.

Also, note that the loss function is dependent on user defined layers, whereas the optimizer is not. An optimizer is not coupled to a model in the same way a loss function is. A templated model can't know in advance what layers to pass to a loss function that is external to the code, whereas an optimizer never needs to know that information anyway.

nkashy1 replied:
It does make sense to try different loss functions, and this is a pretty important step in tuning a model.

The way I'm thinking about it is that I would like to specify (and parametrize) loss functions from the command line when I train a model. Currently, `aleph` is not aware of losses which means I have to extend the CLI to be aware of these parameters if I need the functionality.

`aleph` doesn't fill in the model code anyway at the moment. If we specify the loss function input and output shapes in the `losses.py` file, then we can add them in a comment in the generated model file for the user to fill in.

phildow replied:
https://github.com/doc-ai/aleph/issues/28

nkashy1 replied:
Yep, this would be an important feature, and I would like for this kind of stuff to appear when I run `aleph ... --help`.

It would also allow us to write useful loss abstractions (e.g. compound losses by averaging, etc.)

nkashy1 commented on README.md:
```diff
@@ -1,6 +1,596 @@
+
+#### Built-In Parameters
+
+Let's look at some of the built-in parameters you may override:
```
We should extend train to accept a full training schedule (for example, if we want to schedule learning rate decay).

Look at the [tf.Estimator `train_and_evaluate`](https://www.tensorflow.org/api_docs/python/tf/estimator/train_and_evaluate) method for inspiration here. Some of their design choices (e.g. `EvalSpec` `start_delay_secs` and `throttle_secs` parameters) are made to support distributed training, which we probably shouldn't concern ourselves with at the moment, but most of it comes from pretty standard ML workflows.

nkashy1 replied:
User-defined parameters sound good for now, although we should make things more formal for optimizers (and other resources) so that the parameters show up in `aleph <blah blah blah> --help`.

That would make a good issue for the future, though, and definitely does not block this PR.

phildow replied:
https://github.com/doc-ai/aleph/issues/32 AND https://github.com/doc-ai/aleph/issues/21 AND https://github.com/doc-ai/aleph/issues/20

nkashy1 commented on README.md:
```diff
@@ -1,6 +1,596 @@
+
+You may specify your own user defined parameters. These are passed directly to the train, export, and predict scripts, and any parameters prefixed with the `model` keypath are sent to your model's `model_fn` in the `params` argument.
+
+For example, let's say you're training an image model and using transfer learning with a [TensorFlow Hub](https://tfhub.dev/) module, but you want to be able to parameterize the module you use. Pass a user defined parameter to the train command with a `model` keypath prefix:
```
What is a user-defined parameter? Shouldn't the parameters correspond to dictionary keys?

nkashy1 commented on aleph/cli/init/generator.py:
```diff
@@ -0,0 +1,129 @@
+
+  return os.path.join(templates_path(), 'export.py.jinja2')
+
+def predict_template_path():
```
This is messy. I would introduce a single function:
```
FILE_MAPPER = {
    'predict': 'predict.py.jinja2',
    ...
}

def template_path(template):
    return os.path.join(templates_path(), FILE_MAPPER[template])
```

`FILE_MAPPER.keys()` enumerates over all the possible template types.

nkashy1 replied:
I would also expose the keys as Python objects. So:

```
PREDICT = 'predict'

FILE_MAPPER = {
    PREDICT: 'predict.py.jinja2',
    ...
}
```

nkashy1 replied:
This means that anyone that has good editor/IDE support for Python will be able to autocomplete easily for objects from this module. (It also reduces errors in general.)

nkashy1 replied:
:+1: This would be a good issue (to tackle in separate PR)

phildow replied:
https://github.com/doc-ai/aleph/issues/29

nkashy1 commented on aleph/cli/init/templates/requirements.txt.jinja2:
```diff
@@ -11,10 +11,11 @@ Markdown==3.0.1
 protobuf==3.6.1
+scipy==1.1.0
 six==1.11.0
 tensorboard==1.12.0
```
What if I want to use `tensorflow-gpu`?

nkashy1 replied:
I recommend taking optional arguments to `aleph init` which sets these values in `requirements.txt` when it is generated from template. You can specify defaults somewhere.

phildow replied:
Ah, perhaps we define two requirements.txt files for the project: requirements.txt and requirements-gpu.txt, while also supporting that gpu param at initialization.

This raises a larger question that I've been considering about the typical engineer's workflow. Coming from a django or rails background, you typically have two environments, and you circle between them. So I'll be developing locally in a dev environment, with a local database, and then push to prod, which runs in a prod environment with a production database. When I make changes, I make them in dev locally then push them to prod. Typically cycle.

I see something similar with the aleph workflow: develop locally, perhaps on a device with no gpu support, perhaps with a dev dataset and a low number of training epochs, in order to confirm that model code is correct, then train "in prod" on a device with gpu support on the full dataset. 

Right now there's no way to support splitting the dataset into "dev" and "prod" distributions, although I'm already experiencing a desire to do that with the dataset with the 45GB dataset we just uploaded to a GCS bucket. Locally I don't want to pull from GCS to run training when I could have a gig of "dev" data on my device, and without gpu support, as I test my models, but "in prod" I want to run with gpu support on the full training set.

phildow replied:
There's another issue here: you've already installed aleph before you set up your aleph project, see comments where you begin "From a user experience point of view, would it be possible to specify the working directory for the project one is initializing?" And the aleph package already has tensorflow listed as a dependency and not tensorflow-gpu. If you're installing from a requirements.txt file, say, after uploading the project to a cloud machine, it's not an issue, but if you're on a local device and have already installed aleph, you'd need to uninstall tensorflow and then pip install tensorflow-gpu

nkashy1 replied:
Exactly. (Even with versions.) I would make this an issue and tackle it in a separate PR.

nkashy1 replied:
My earlier comment was in response to your third to last comment.

I would make the issue more open-ended. Listing all these concerns and invite others for discussion (e.g. @axelsly, @kthakore ). We don't have to solve the problem in this PR.

phildow replied:
I think at a minimum every project has both a requirements.txt and requirements-gpu.txt fil

phildow replied:
https://github.com/doc-ai/aleph/issues/30 AND https://github.com/doc-ai/aleph/issues/31

nkashy1 commented on aleph/cli/models/generator.py:
```diff
@@ -0,0 +1,118 @@
+
+  exit_if_not_aleph_project()
+  
+  if args.subcommand == 'list':
```
See comment on earlier pull request re: defining subcommand handlers directly when you register the subparser: https://docs.python.org/3/library/argparse.html#argparse.ArgumentParser.add_subparsers (CTRL+F for `# sub-command functions`).

phildow replied:
I could make an issue for this, I'll have to look into how to handle this with subparsers, and it makes the importer messier. I also like having a single entry point in aleph.command_line_interface where i can catch exceptions.

nkashy1 replied:
If it is an issue, it's a low priority one. It's pretty easy to reason about this code, and that's all anyone can ask for really.

phildow replied:
https://github.com/doc-ai/aleph/issues/33

nkashy1 commented:
@phildow : Should the `aleph/aleph_cli` directory be deleted?

nkashy1 commented:
Re: Test directory structure -- note that the existence of the Stackoverflow question means that this is not the Python-friendliest organization of code. It means that you have to import packages/modules into your test files that will throw errors unless you run tests using the unittest CLI. I don't particularly like the restriction. Do you feel strongly about test organization?

It's pretty standard to put test files along the code files.

nkashy1 commented:
Re: `aleph/command_line_interface.py`

You can use Python's `__import__` method to perform dynamic imports:
```
command_line_interface = __import__('aleph.command_line_interface')
```

Using this in that file would make it cleaner (and would remove the requirement of having to update that file every time we made a change any arbitrary number of levels down the resource definitions).

phildow commented:
1. Should the aleph/aleph_cli directory be deleted?

Yes it should be, but I'm not seeing that in this branch.

2. Re: Test directory structure

I do feel strongly about this. Most of what I've read does suggest that more complex projects put tests in a submodule somewhere, even if that's in the module itself. So at the very least:

```
aleph/
  cli/
    datasets/
      arg_parser.py
      tests/
        test_arg_parser.py
```

My own personal preference is to have all tests in a separate top level submodule entirely, as project complexity is only going to grow, but i'm also not coming from a python background, and python packaging/modules/importing is one of the few things about the language that I particularly dislike.

3. Re: aleph/command_line_interface.py

I'll have a look at cleaning that up.

phildow commented:
How do you want to handle this PR? I could address all/most of these comments or open up issues for some of them?

This PR did get out of hand. I'd definitely like smaller PRs, but there was a lot of stacking here. And I would make a request for faster turnaround on reviews. :) I'm pushing pretty hard on this code, and I know you're stretched on multiple projects.

nkashy1 commented:
@phildow 

re: aleph_cli directory
That just existed on my machine (strangely). I removed it and it did not affect my git status.

re: test structure
My preference is mild at best. I would say look at a bunch of open source Python projects (NOT TensorFlow, though :) ) and let's do what they do. It seems as if PyTorch prefers your organization. I'm cool with it. Make an issue and I will port over the existing test to the new setup.

re: command_line_interface.py cleanup
Would make a good issue.

re: quicker turnaround
I've wrapped up most of the other projects, so will be spending more time thinking only about AI, so turnaround on reviews should be MUCH quicker now.
